cilium: add IP option packet tracing feature

Bigdelle opened the pull request:
This PR addresses [this design CFP](https://github.com/cilium/design-cfps/pull/44).

This change introduces packet tracing via trace IDs embedded in IP options, observable via `cilium monitor` and `hubble`

Bigdelle committed:
feat: add IPv4 packet tracing via IP options.

- Added feature-gate `enable-ip-option-tracing` to set IP option tracing.
- Implemented helper functions to parse IPv4 options.
- Added logic to save parsed IPv4 options into a per-CPU array map.

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

Bigdelle committed:
feat: add IP option tracing parsing in ingress programs

- Implemented parsing logic for IP options in ingress programs.
- Store extracted IP options into per-CPU array map.

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

Bigdelle committed:
feat: integrate IP option trace ID into event messages

- Add 'ip_trace_id' field into drop and trace structs.
- Modified DropNotify and TraceNotify structs to include IPTraceID field.

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

sypakine commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,199 @@
+#include "common.h"
+#include "ipv6.h"
+
+/* Length of the first supported IPv4 trace_id option (4 bytes). */
```
Let's clarify that this is 2 bytes (fixed) + X bytes (variable) and what bit length is supported.

sypakine commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,199 @@
+#include "ipv6.h"
+
+/* Length of the first supported IPv4 trace_id option (4 bytes). */
+#define IPV4_OPT1_LEN 4
```
The # in these doesn't have any meaning. Can we use the bit or word length of the trace value?

sypakine commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,199 @@
+		if (optlen != IPV4_OPT1_LEN && optlen != IPV4_OPT2_LEN && optlen != IPV4_OPT3_LEN)
+			return TRACE_ID_INVALID;
+
+		if (optlen == 4) {
```
See if switch looks cleaner?

```
switch(optlen) {
case IPV4_OPT16_LEN:
  __s16 temp;

  if (ctx_load_bytes(ctx, offset + 2, &temp, sizeof(temp)) < 0)
    return TRACE_ID_ERROR;

  temp = bpf_ntohs(temp);
  if (temp <= 0)
    return TRACE_ID_INVALID;
  trace_id = temp;
  return trace_id
case IPV4_OPT32_LEN:
  ...
case IPV4_OPT64_LEN:
  ...
default:
  return TRACE_ID_ERROR
}
```

Bigdelle replied:
Updated and fixed.

sypakine commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,199 @@
+ * - Returns TRACE_ID_NO_FAMILY if the packet is not IPv4.
+ * - Returns TRACE_ID_SKIP_IPV6 if the packet is IPv6.
+ */
+static __always_inline __s64 trace_id_from_ctx(struct __ctx_buff *ctx, __u8 ip_opt_type_value)
```
Is there a reason we cannot use an unsigned number?

Bigdelle replied:
We output negative MACRO values for various reasons (invalid trace id, trace id not found, etc). So if we output a negative number in an unsigned function, we'll have unexpected behavior (I believe in C it wraps around).

sypakine commented on daemon/cmd/daemon_main.go:
```diff
@@ -1143,6 +1143,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.StringSlice(option.NodeLabels, []string{}, "List of label prefixes used to determine identity of a node (used only when enable-node-selector-labels is enabled)")
 	option.BindEnv(vp, option.NodeLabels)
 
+	flags.Int(option.EnableIPOptionTracing, 0, "Enable packet IP tracing. Set this option to specify the IP options field of IPv4 packet from which to read tracing information; a value of 0 disables IP tracing.")
```
`GetUint` and you don't need to specify negative numbers are unsupported.

and the config.go usage below cannot be silently ignored.

Bigdelle replied:
In config.go we only set the macro if the `<type>` set is positive because we use this as the macro to do parsing if it is defined.

If we change the config such that it always defines the macro with the value passed in by the user, this means that we would have to change how we do the parsing:
```
#ifdef ENABLE_PACKET_IP_TRACING
	check_and_store_ip_trace_id(ctx, ENABLE_PACKET_IP_TRACING);
#endif
```

We can do that, but this would just remove the check from the configuration and put it into the bpf code instead, which I don't know if there is much value in doing.

sypakine replied:
As a user, what if I set `-136`? Current implementation ignores that intent and treats it as `0`. Changing this to use `Uint` would cause the agent to error out and report that an invalid type is used for the option.

Further, this should be `Uint8` to constrain the input to 0-255 (the IPv4 option type range)

https://github.com/spf13/pflag/blob/d5e0c0615acee7028e1e2740a11102313be88de1/uint8.go#L81

Bigdelle replied:
Agreed. Code has been changed to reflect this.

sypakine reviewed (commented):
We should only render the code if the feature is enabled.

Bigdelle committed:
feat: add `IPTraceID` into flow struct

- Modified flow.proto to include `IPTraceID` struct
- Add IPTraceID field to Flow and FlowFilter struct
- Add parsing to decode `IPTraceID` into Flow

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

Bigdelle committed:
feat: add filter support for `IPTraceID`

- Implemented filtering logic for `IPTraceID`
- Create testing file for new filter
- Add filter to supported DefaultFilters

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

Bigdelle committed:
feat: add IPTraceID filter to hubble

- Add filter flag `--ip-trace-id` to hubble.
- Add `IPTraceID` output to hubble printer.
- Update testing for `observe help` command.

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

Bigdelle committed:
feat: integrated `ip_trace_id` with Hubble printer

- Implemented feature to print `ip_trace_id` if it exists.

Signed-off-by: Ben Bigdelle <bigdelle@google.com>

kaworu commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,204 @@
+
+/* The minimum value for IHL which corresponds to a packet with no options.
+ *
+ * A standard IP packet header has 20 bytes and the IHL is the number of 32 byte
+ * words.
```
typo
```suggestion
 * A standard IP packet header has 20 bytes and the IHL is the number of 32-bit
 * words.
```

kaworu commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,204 @@
+ */
+#define IHL_WITH_NO_OPTS 5
+
+/* Signifies that options were parsed correctly but no trace ID was found. */
+#define TRACE_ID_NOT_FOUND 0
```
As far as I understand there is no way for callers of `trace_id_from_ip4()` to distinguish between `TRACE_ID_NOT_FOUND` and success. Functionally `TRACE_ID_NOT_FOUND` like other errors will end up setting the trace id to `0` in the map, but it is still confusing IMO.

sypakine replied:
Agreed. Bumping all these return codes and adding `TRACE_ID_OK` as 0.

kaworu commented on bpf/lib/ip_options.h:
Any reason to use signed values for trace id in this file?

kaworu commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,204 @@
+ */
+#define TRACE_ID_SKIP_IPV6 -100
+
+/* Converts a 64-bit integer from network byte order to host byte order. */
+static __always_inline __u64 ntohll(__u64 value)
+{
+	return ((value & 0x00000000000000FFULL) << 56) |
+		   ((value & 0x000000000000FF00ULL) << 40) |
+		   ((value & 0x0000000000FF0000ULL) << 24) |
+		   ((value & 0x00000000FF000000ULL) << 8)  |
+		   ((value & 0x000000FF00000000ULL) >> 8)  |
+		   ((value & 0x0000FF0000000000ULL) >> 24) |
+		   ((value & 0x00FF000000000000ULL) >> 40) |
+		   ((value & 0xFF00000000000000ULL) >> 56);
+}
```
Looks to me that we're swapping regardless of the host byte order. Why not [`bpf_cpu_to_be64`](https://github.com/cilium/cilium/blob/7df597a49bbd9092550307a075e93ee1f89a68ca/bpf/lib/endian.h#L74-L76) instead?

sypakine replied:
Thanks for the great suggestion -- will test and removing the associated tests.

kaworu commented on daemon/cmd/daemon_main.go:
```diff
@@ -1141,6 +1141,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.StringSlice(option.NodeLabels, []string{}, "List of label prefixes used to determine identity of a node (used only when enable-node-selector-labels is enabled)")
 	option.BindEnv(vp, option.NodeLabels)
 
+	flags.Uint8(option.EnableIPOptionTracing, 0, "Enable packet IP tracing. Set this option to specify the IP options field of IPv4 packet from which to read tracing information; a value of 0 disables IP tracing.")
```
Should we somehow prevent users to set `option.EnableIPOptionTracing=1` as we always skip `IPOPT_NOOP`?

sypakine replied:
Good catch. Adding validation.

kaworu commented on api/v1/flow/flow.proto:
```diff
@@ -82,6 +82,9 @@ message Flow {
     // Cilium datapath trace reason info.
     TraceReason trace_reason = 36;
 
+    // IPTraceID relates to the trace ID in the IP options of a packet.
+    IPTraceID ip_trace_id = 38;
```
Should this be in the IP message instead?

kaworu commented on bpf/lib/drop.h:
```diff
@@ -53,6 +56,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_DROP_NOTIFY)
 int __send_drop_notify(struct __ctx_buff *ctx)
 {
 	/* Mask needed to calm verifier. */
+	__u64 ip_trace_id = load_ip_trace_id();
 	__u32 error = ctx_load_meta(ctx, 2) & 0xFFFFFFFF;
```
keep the comment and mask line together

```suggestion
	__u64 ip_trace_id = load_ip_trace_id();
	/* Mask needed to calm verifier. */
	__u32 error = ctx_load_meta(ctx, 2) & 0xFFFFFFFF;
```

kaworu commented on bpf/lib/trace.h:
```diff
@@ -319,6 +326,7 @@ send_trace_notify6(struct __ctx_buff *ctx, enum trace_point obs_point,
 		.reason		= reason,
 		.ifindex	= ifindex,
 		.ipv6		= 1,
+		.ip_trace_id = ip_trace_id,
```
I don't think we need this for `send_trace_notify6` as long as it's an IPv4 only feature.

sypakine replied:
yup, will remove.

kaworu commented on pkg/hubble/parser/threefour/parser.go:
```diff
@@ -459,6 +461,19 @@ func decodeTraceReason(tn *monitor.TraceNotify) pb.TraceReason {
 }
 
+func decodeIpTraceId(tn *monitor.TraceNotify) *pb.IPTraceID {
+	if tn == nil {
+		return &pb.IPTraceID{
+			TraceId:      0,
+			IpOptionType: 0,
+		}
+	}
+	return &pb.IPTraceID{
+		TraceId:      tn.GetIPTraceID(),
+		IpOptionType: uint32(option.Config.EnableIPOptionTracing),
+	}
```
return `nil` when possible, that way the json encoder will skip the property and emit `{flow:{…}}` instead of an empty trace id `{flow:{…, ip_trace_id:{}}}`
```suggestion
	if tn == nil || option.Config.EnableIPOptionTracing == 0 {
		return nil
	}
	return &pb.IPTraceID{
		TraceId:      tn.GetIPTraceID(),
		IpOptionType: uint32(option.Config.EnableIPOptionTracing),
	}
```

kaworu commented on hubble/cmd/observe/flows_filter.go:
```diff
@@ -530,6 +531,15 @@ func (of *flowFilter) set(f *filterTracker, name, val string, track bool) error
+	case "ip-trace-id":
+		traceID, err := strconv.ParseUint(val, 16, 64)
+		if err != nil {
+			return err
```
```suggestion
			return fmt.Errorf("invalid --ip-trace-id value: %w", err)
```

kaworu commented on hubble/pkg/printer/printer.go:
```diff
@@ -372,10 +379,16 @@ func (p *Printer) WriteProtoFlow(res *observerpb.GetFlowsResponse) error {
 			arrow = "<-"
 		}
+
+		if f.GetIpTraceId().GetTraceId() != 0 {
+			traceID = p.GetIpTraceID(f)
+		}
+
 		_, err := fmt.Fprintf(p.opts.w,
-			"%s%s: %s %s %s %s %s %s %s (%s)\n",
+			"%s%s%s: %s %s %s %s %s %s %s (%s)\n",
 			fmtTimestamp(p.opts.timeFormat, f.GetTime()),
 			node,
+			traceID,
```
Not sure we want this in the compact output cc @michi-covalent

kaworu commented on pkg/monitor/datapath_trace.go:
```diff
@@ -22,9 +22,9 @@ const (
 	// traceNotifyV0Len is the amount of packet data provided in a trace notification v0.
-	traceNotifyV0Len = 32
+	traceNotifyV0Len = 40
 	// traceNotifyV1Len is the amount of packet data provided in a trace notification v1.
-	traceNotifyV1Len = 48
+	traceNotifyV1Len = 56
```
Can we do this @gandro?

gandro replied:
Oh good catch. Yeah, I believe this is a breaking change, we can't do this.

These structs are written to the perf events ring buffer by the datapath and read by user space. The events map is (to my knowledge) not replaced during upgrades (or downgrades for that matter), and so it will happen that the buffer will be full with events written with the shorter length, but userspace expects them to have the new (longer length) and thus will lead to invalid data being read.

I think we will have to introduce a version 2 and version 3 of the trace notify struct to achieve these modifications, and ensure the agent can deal with all 4 versions. Once we no longer need to support upgrading from Cilium v1.16, then we can get rid of v0 and v1.

sypakine replied:
i'm kinda hung on this; it seems like an interface for this would best address the versioning issue, with the current struct working to provide the base notification information.

I'm going to experiment with that change.

kaworu reviewed (changes requested):
Thanks for the PR @Bigdelle, some comments.

As it stand it feels like the IP option parsing would be simpler if implemented by the Hubble threefour parser, but I guess some cases (e.g. encap) and the CFP future milestones require manipulating the trace id from the datapath.

pchaigno commented on bpf/lib/ip_options.h:
```diff
@@ -0,0 +1,204 @@
+		}
+
+		if (optlen != OPT16_LEN && optlen != OPT32_LEN && optlen != OPT64_LEN)
+			return TRACE_ID_INVALID;
```
This looks unnecessary given the check and error with `TRACE_ID_UNSUPPORTED_LENGTH_ERROR` below.

sypakine replied:
Yup, removing and updating tests.

pchaigno commented on bpf/lib/trace_helpers.h:
```diff
@@ -0,0 +1,74 @@
+	__u64 *__cache = map_lookup_elem(&trace_id_map, &__z);
+
+	if (__cache)
+		*__cache = trace_id;
```
Why not just use `return map_update_elem(&trace_id_map, &__z, &trace_id, BPF_ANY);`?

sypakine replied:
Thanks, will make that change.

pchaigno commented on bpf/lib/trace_helpers.h:
```diff
@@ -0,0 +1,74 @@
+static __always_inline __u64 load_ip_trace_id(void)
+{
+	return bpf_trace_id_get();
+}
```
What is the purpose of this function? It has the same prototype as `bpf_trace_id_get` and just calls it.

sypakine replied:
Thanks, removing.

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -793,6 +793,9 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	cDefinesMap["CLUSTER_ID_MAX"] = fmt.Sprintf("%d", option.Config.MaxConnectedClusters)
 
+	// Write ENABLE_PACKET_IP_TRACING macro from EnablePacketIPTracing value
+	cDefinesMap["ENABLE_PACKET_IP_TRACING"] = strconv.Itoa(int(option.Config.EnableIPOptionTracing))
```
I forgot that in my review: you will probably want to add coverage for this in complexity tests (`bpf/complexity-tests/`). There's quite a bit of parsing logic under this macro, so you probably want to ensure it's covered so that you don't hit a verifier error when you enable it.

sypakine replied:
Good call. when I finish addressing the other comments in this PR, i'll get to that.
